File: mobile/lib/api/plans.ts
import { handler } from "@/lib/api/api";
import { cachedQuery, invalidateCachedQuery, prefetchQuery } from "@/lib/api/query-cache";
import type { UserTier } from "@/lib/types/auth";

export interface MobilePlanCatalogItem {
  tier: UserTier;
  name: string;
  guest_limit: number;
  billing_period: "P1M";
  description: string;
}

export interface PlanAvailability {
  google_play: boolean;
  app_store: boolean;
}

export interface PlanCatalogResponse {
  plans: MobilePlanCatalogItem[];
  availability: PlanAvailability;
}

export interface SubscriptionStatus {
  tier: UserTier;
  guest_limit: number;
  active: boolean;
  status: "free" | "active" | "cancelled" | "billing_issue" | "paused" | "expired";
  billing_period: "P1M" | null;
  started_at: string | null;
  expires_at: string | null;
  product_id: string | null;
  store: "google_play" | "app_store" | null;
  auto_renews: boolean;
  availability: PlanAvailability;
}

export interface GuestLimitStatus {
  allowed: boolean;
  current: number;
  limit: number;
  tier: UserTier;
}

export function fetchPlanCatalog(): Promise<PlanCatalogResponse> {
  return cachedQuery(
    "plan-catalog",
    () => handler<PlanCatalogResponse>("/api/v1/plans/catalog", { auth: false }),
    { ttlMs: 300_000 },
  );
}

export function fetchSubscriptionStatus(options: { force?: boolean } = {}): Promise<SubscriptionStatus> {
  return cachedQuery(
    "subscription-status",
    () => handler<SubscriptionStatus>("/api/v1/plans/status"),
    { ttlMs: 30_000, force: options.force },
  );
}

export async function syncPlanSubscription(): Promise<SubscriptionStatus> {
  const status = await handler<SubscriptionStatus>("/api/v1/plans/sync", { method: "POST" });
  invalidateCachedQuery("subscription-status");
  invalidateCachedQuery("guest-limit:");
  return status;
}

export function checkGuestLimit(
  eventId: string,
  options: { force?: boolean } = {},
): Promise<GuestLimitStatus> {
  return cachedQuery(
    `guest-limit:${eventId}`,
    () => handler<GuestLimitStatus>(
      `/api/v1/guests/events/${encodeURIComponent(eventId)}/limit`,
    ),
    { ttlMs: 15_000, force: options.force },
  );
}

export function invalidateGuestLimit(eventId?: string) {
  invalidateCachedQuery(eventId ? `guest-limit:${eventId}` : "guest-limit:");
}

export function prefetchWorkspacePlanData(eventId: string, includeGuestUsage = true) {
  prefetchQuery(fetchPlanCatalog());
  prefetchQuery(fetchSubscriptionStatus());
  if (includeGuestUsage) prefetchQuery(checkGuestLimit(eventId));
}

File: app/models/schemas/plans.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.user import UserTier, UserTierStatus, UserTierStore


class PlanAvailability(BaseModel):
    google_play: bool
    app_store: bool


class PlanCatalogItem(BaseModel):
    tier: UserTier
    name: str
    guest_limit: int
    billing_period: str = "P1M"
    description: str


class PlanCatalogResponse(BaseModel):
    plans: list[PlanCatalogItem]
    availability: PlanAvailability


class SubscriptionStatusResponse(BaseModel):
    tier: UserTier
    guest_limit: int
    active: bool
    status: UserTierStatus
    billing_period: str | None = None
    started_at: datetime | None = None
    expires_at: datetime | None = None
    product_id: str | None = None
    store: UserTierStore | None = None
    auto_renews: bool = False
    availability: PlanAvailability


class RevenueCatWebhookEvent(BaseModel):
    id: str = Field(..., min_length=1)
    type: str = Field(..., min_length=1)
    app_user_id: str | None = None
    original_app_user_id: str | None = None
    aliases: list[str] = Field(default_factory=list)
    product_id: str | None = None
    store: str | None = None
    expiration_at_ms: int | None = None
    entitlement_ids: list[str] = Field(default_factory=list)
    transferred_from: list[str] = Field(default_factory=list)
    transferred_to: list[str] = Field(default_factory=list)


class RevenueCatWebhookRequest(BaseModel):
    api_version: str | None = None
    event: RevenueCatWebhookEvent
    raw_payload: dict[str, Any] = Field(default_factory=dict, exclude=True)


class WebhookAcceptedResponse(BaseModel):
    status: str
    duplicate: bool = False
